File: geeksgforgeeks/challenge5_coinChange/coinChange.py
# Created by Avinash on 4th August, 2017
def process():
	amount = int(input("Enter the amount : "))
	coins = int(input("Enter the number of denominations : "))
	denominations = [[0] for i in range(coins)]
	print("Enter the denominations : ")
	for i in range(coins) : 
		denominations[i] = int(input())
	coinChangeWays = coinChange(amount,coins,denominations)
	print("The total number of ways to make the change using coins is : %-5d" %coinChangeWays)

def coinChange(n, m, d) :
	cache = [[0] * (n + 1) for i in range(m)]
	for i in range(m) :
		cache[i][0] = 1
	for i in range(m) :
		for j in range(1,n + 1) :
			include_coin = cache[i][j - d[i]] if (j - d[i] >= 0) else 0
			exclude_coin = cache[i - 1][j] if i >= 1 else 0
			cache[i][j] = include_coin + exclude_coin
	return cache[m-1][n]

if __name__=="__main__":
	process()
		
	

File: geeksgforgeeks/challenge5_coinChange/CoinChange.java
/*
     Created by Avinash on 4th August, 2017
 */
import java.io.*;
import java.util.*;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;
public class CoinChange 
{
	public static int coinChangeWays;
	public static void main(String[] args) 
	{
        	BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
		int amount;
		int coins, denominations[];
		try
		{    
	        	System.out.print("Enter the amount : ");
	        	amount = Integer.parseInt(read.readLine());
	        	System.out.print("Enter the number of denominations : ");
	        	coins = Integer.parseInt(read.readLine());
	        	denominations = new int[coins];
	        	System.out.print("Enter the denominations : ");
	        	for(int i = 0; i < coins; i++)
	        		denominations[i] = Integer.parseInt(read.readLine());
			Helper link = new Helper();
			coinChangeWays = link.coinChange(amount, coins, denominations);
			
		}
		catch (IOException e)
		{
			e.printStackTrace();
		}
                System.out.println("The number of ways to get the change is : "+coinChangeWays);
	}

}
class Helper
{

        public int coinChange(int n, int m, int[] d)
	{
		int[][] t = new int[m][n+1];
		int include_coin, exclude_coin;
 		for(int i = 0; i < m; i++)
 			t[i][0] = 1;
 		for(int i = 0; i < m; i++)
 		{
 			for( int j = 1; j < n+1; j++)
 			{
 				include_coin  = ( j - d[i] >= 0) ? t[i][j - d[i]] : 0;
 				exclude_coin = (i >= 1) ? t[i-1][j] : 0;
 				t[i][j] = include_coin + exclude_coin;
 			}
 		}
 		return t[m-1][n];	
	}				
}
         
